cachix/src/Cachix/Client/PushQueue.hs

```diff
@@ -0,0 +1,102 @@
+{- Implements a queue with the following properties:
+
+- waits for queue to be fully pushed when exiting using ctrl-c (SIGINT)
+- allows stopping the producer
+-
+- avoid duplicate pushing of the same store paths
+
+To safetly exit on demand, signal SIGINT.
+-}
+module Cachix.Client.PushQueue
+  ( startWorkers,
+    Queue,
+  )
+where
+
+import qualified Cachix.Client.Push as Push
+import Cachix.Client.Retry (retryAll)
+import Control.Concurrent.Async
+import Control.Concurrent.STM (TVar, modifyTVar', newTVarIO, readTVarIO)
+import qualified Control.Concurrent.STM.TBQueue as TBQueue
+import qualified Data.Set as S
+import Protolude
+import qualified System.Posix.Signals as Signals
+
+type StorePath = Text
+
+type Queue = TBQueue.TBQueue StorePath
+
+data PushWorkerState
+  = PushWorkerState
+      { pushQueue :: Queue,
+        inProgress :: TVar Int
+      }
+
+data QueryWorkerState
+  = QueryWorkerState
+      { queryQueue :: Queue,
+        alreadyQueued :: S.Set StorePath
+      }
+
+worker :: Push.PushCache IO () -> PushWorkerState -> IO ()
+worker pushCache workerState = forever $ do
+  storePath <- atomically $ TBQueue.readTBQueue $ pushQueue workerState
+  bracket_ (inProgresModify (+ 1)) (inProgresModify (\x -> x - 1))
+    $ retryAll
+    $ Push.uploadStorePath pushCache storePath
+  where
+    inProgresModify f =
+      atomically $ modifyTVar' (inProgress workerState) f
+
+-- NOTE: producer is responsible for signaling SIGINT upon termination
+-- NOTE: producer should return an `IO ()` that should be a blocking operation for terminating it
+startWorkers :: Int -> (Queue -> IO (IO ())) -> Push.PushCache IO () -> IO ()
+startWorkers numWorkers mkProducer pushCache = do
+  -- start query worker
+  queryQueue <- atomically $ TBQueue.newTBQueue 10000
+  pushQueue <- atomically $ TBQueue.newTBQueue 10000
+  queryWorker <- async $ queryLoop (QueryWorkerState queryQueue S.empty) pushQueue pushCache
+  -- start push workers
+  stopProducerCallback <- mkProducer queryQueue
+  progress <- newTVarIO 0
+  let workerState = PushWorkerState pushQueue progress
+  consumer <- async $ replicateConcurrently_ numWorkers $ worker pushCache workerState
+  void $ Signals.installHandler Signals.sigINT (Signals.CatchOnce (exitOnceQueueIsEmpty stopProducerCallback consumer workerState)) Nothing
+  waitEither_ consumer queryWorker
+
+queryLoop :: QueryWorkerState -> Queue -> Push.PushCache IO () -> IO ()
+queryLoop workerState pushQueue pushCache = do
+  storePaths <- atomically $ do
+    -- this blocks until item is available and doesn't remove it from the queue
+    _ <- TBQueue.peekTBQueue (queryQueue workerState)
+    TBQueue.flushTBQueue (queryQueue workerState)
```

Comment:
Equivalent to flush + `guard` not empty, fwiw

Reply:
Flush says it never retries, so it would then need handle the loop

Reply:
The `guard` would retry it.